File: template/java/sevenEleven/BeverageMaker.java
/**
 * 樣板方法模式 (Template Method Pattern)
 * 
 * 核心概念：
 * 1. 在一個方法中定義一個演算法的骨架（SOP），而將一些步驟延遲到子類別中。
 * 2. 樣板方法使得子類別可以在不改變演算法結構的情況下，重新定義演算法的某些步驟。
 * 3. 透過「掛鉤」(Hook) 機制，讓子類別能有選擇性地影響演算法流程。
 * 
 * 
 * BeverageMaker 飲料自動製作機
 * 
 * 抽象超類別 (Abstract Superclass)
 * - Beverage 飲料
 * 
 * 具體類別 (Concrete Class)
 * - Coffee 咖啡
 * - Tea 茶
 * - MilkTea 珍珠奶茶
 */

// 實作展示：7-11 飲料自動製作機
public class BeverageMaker {
  public static void main(String[] args) {
    System.out.println("--- 製作 Coffee ---");
    Beverage coffee = new Coffee();
    coffee.prepareRecipe();

    System.out.println("\n--- 製作 Tea (不加檸檬) ---");
    Beverage tea = new Tea(false); // 透過 Hook 控制不加料
    tea.prepareRecipe();

    System.out.println("\n--- 製作 MilkTea ---");
    Beverage milkTea = new MilkTea();
    milkTea.prepareRecipe();
  }
}

/**
 * [抽象超類別]：定義演算法骨架
 */
abstract class Beverage {

  /**
   * 樣板方法 (Template Method)
   * 宣告為 final 是為了防止子類別修改這個演算法 SOP 結構。
   */
  public final void prepareRecipe() {
    boilWater(); // 步驟 1：煮開水 (共用)
    brew(); // 步驟 2：沖泡 (子類別實作)
    pourInCup(); // 步驟 3：倒入杯子 (共用)

    // 使用「掛鉤」(Hook) 來決定是否要執行步驟 4
    if (customerWantsCondiments()) {
      addCondiments(); // 步驟 4：加配料 (子類別實作)
    }
  }

  /** 抽象方法：由子類別決定如何沖泡 */
  abstract void brew();

  /** 抽象方法：由子類別決定加什麼配料 */
  abstract void addCondiments();

  /** 共用方法：所有飲料都要煮開水 */
  private void boilWater() {
    System.out.println("[Beverage] 正在煮沸開水...");
  }

  /** 共用方法：所有飲料都要倒進杯子 */
  private void pourInCup() {
    System.out.println("[Beverage] 將飲料倒入杯中...");
  }

  /**
   * 掛鉤 (Hook)：子類別可以選擇性地覆寫它
   * 預設為 true (要加料)。
   */
  boolean customerWantsCondiments() {
    return true;
  }
}

/**
 * [具體類別]：咖啡
 */
class Coffee extends Beverage {
  @Override
  void brew() {
    System.out.println("[Coffee] 透過濾網滴濾咖啡");
  }

  @Override
  void addCondiments() {
    System.out.println("[Coffee] 加入糖與牛奶");
  }
}

/**
 * [具體類別]：茶
 */
class Tea extends Beverage {
  private boolean wantsCondiments = true;

  public Tea() {
  }

  public Tea(boolean wantsCondiments) {
    this.wantsCondiments = wantsCondiments;
  }

  @Override
  void brew() {
    System.out.println("[Tea] 浸泡茶包");
  }

  @Override
  void addCondiments() {
    System.out.println("[Tea] 加入新鮮檸檬片");
  }

  /** 覆寫掛鉤：根據建構子決定是否要加料 */
  @Override
  boolean customerWantsCondiments() {
    return this.wantsCondiments;
  }
}

/**
 * [具體類別]：珍珠奶茶 (展示易擴充性)
 */
class MilkTea extends Beverage {
  @Override
  void brew() {
    System.out.println("[MilkTea] 沖泡濃郁紅茶並加入牛奶");
  }

  @Override
  void addCondiments() {
    System.out.println("[MilkTea] 加入大顆珍珠（波霸）");
  }
}

/**
 * 總結：
 * 1. 易懂：超類別掌控了整個 SOP，子類別只需專注於自己的特色步驟。
 * 2. 易擴充：如果要新增「熱可可」，只需繼承 Beverage 並實作 brew() 與 addCondiments()。
 * 3. 程式碼複用：共用的邏輯（煮水、倒杯子）都在超類別，避免重複代碼。
 * 
 * 
 * --- 製作 Coffee ---
 * [Beverage] 正在煮沸開水...
 * [Coffee] 透過濾網滴濾咖啡
 * [Beverage] 將飲料倒入杯中...
 * [Coffee] 加入糖與牛奶
 * 
 * --- 製作 Tea (不加檸檬) ---
 * [Beverage] 正在煮沸開水...
 * [Tea] 浸泡茶包
 * [Beverage] 將飲料倒入杯中...
 * 
 * --- 製作 MilkTea ---
 * [Beverage] 正在煮沸開水...
 * [MilkTea] 沖泡濃郁紅茶並加入牛奶
 * [Beverage] 將飲料倒入杯中...
 * [MilkTea] 加入大顆珍珠（波霸）
 */

File: template/ts/sevenEleven/beverage-maker.ts
/**
 * 樣板方法模式 (Template Method Pattern) - TypeScript 版本
 * 
 * 核心概念：
 * 1. 定義演算法骨架：在超類別中定義一個固定步驟的 SOP。
 * 2. 步驟延遲實作：將具體細節（如沖泡方式、配料）留給子類別去實現。
 * 3. 掛鉤 (Hook)：透過一個預設傳回布林值的方法，讓子類別能「插手」演算法的流程控制。
 */

/**
 * [抽象超類別]：飲料基底
 * 在 TypeScript 中，我們可以使用 abstract class 來強制子類別實作特定方法。
 */
abstract class Beverage {

  /**
   * 樣板方法 (Template Method)
   * 定義了製作飲料的 SOP 流程。
   */
  public prepareRecipe(): void {
    this.boilWater();         // 步驟 1 (共用)
    this.brew();              // 步驟 2 (差異化)
    this.pourInCup();         // 步驟 3 (共用)

    // 透過掛鉤控制步驟 4
    if (this.customerWantsCondiments()) {
      this.addCondiments(); // 步驟 4 (差異化)
    }
  }

  /** 抽象方法：具體如何沖泡，由子類別決定 */
  protected abstract brew(): void;

  /** 抽象方法：具體加什麼配料，由子類別決定 */
  protected abstract addCondiments(): void;

  /** 共用邏輯：煮水 */
  private boilWater(): void {
    console.log("[Beverage] 正在煮沸開水...");
  }

  /** 共用邏輯：倒入杯子 */
  private pourInCup(): void {
    console.log("[Beverage] 將飲料倒入杯中...");
  }

  /** 
   * 掛鉤 (Hook)
   * 提供預設行為，子類別可視需求「覆寫」它來改變流程。
   */
  protected customerWantsCondiments(): boolean {
    return true;
  }
}

// --- 具體實作類別 ---

/**
 * 咖啡類別
 */
class Coffee extends Beverage {
  protected brew(): void {
    console.log("[Coffee] 透過濾網滴濾咖啡豆萃取液");
  }

  protected addCondiments(): void {
    console.log("[Coffee] 加入有機糖與鮮奶油");
  }
}

/**
 * 茶類別 (展示 Hook 的彈性)
 */
class Tea extends Beverage {
  // 透過屬性控制 Hook 行為
  constructor(private readonly wantsLemon: boolean = true) {
    super();
  }

  protected brew(): void {
    console.log("[Tea] 浸泡高山烏龍茶包");
  }

  protected addCondiments(): void {
    console.log("[Tea] 加入新鮮萊姆片");
  }

  /** 覆寫 Hook：根據實例狀態決定是否要加料 */
  protected customerWantsCondiments(): boolean {
    return this.wantsLemon;
  }
}

/**
 * 珍珠奶茶 (易擴充性證明)
 */
class MilkTea extends Beverage {
  protected brew(): void {
    console.log("[MilkTea] 沖泡錫蘭紅茶並混合香醇牛奶");
  }

  protected addCondiments(): void {
    console.log("[MilkTea] 加入Ｑ彈黑糖珍珠");
  }
}

// --- 執行展示 ---

function runBeverageMaker() {
  console.log("--- 1. 製作經典咖啡 ---");
  const coffee = new Coffee();
  coffee.prepareRecipe();

  console.log("\n--- 2. 製作檸檬紅茶 (使用預設加料) ---");
  const lemonTea = new Tea();
  lemonTea.prepareRecipe();

  console.log("\n--- 3. 製作純紅茶 (透過 Hook 拒絕加料) ---");
  const plainTea = new Tea(false);
  plainTea.prepareRecipe();

  console.log("\n--- 4. 製作珍珠奶茶 (新擴充品項) ---");
  const bobaMilkTea = new MilkTea();
  bobaMilkTea.prepareRecipe();
}

runBeverageMaker();

/**
 * TS/JS 版本優點總結：
 * 1. 強型別檢查：使用 abstract 確保開發者不會漏掉必要的步驟。
 * 2. 結構清晰：利用 Class 的繼承特性，將「變」與「不變」的邏輯完美拆分。
 * 3. 易於擴充：想要加新口味？繼承 Beverage 並填空即可，不需修改既有基礎代碼 (符合 OCP)。
 */
